src/v/storage/fs_utils.h

```diff
@@ -63,4 +63,116 @@ struct segment_path {
     }
 };
 
+struct segment_full_path;
+
+struct partition_path {
+    model::ntp ntp;
+    model::revision_id revision_id;
+
+    partition_path(const ntp_config& ntpc) {
+        ntp = ntpc.ntp(), revision_id = ntpc.get_revision();
+    }
+
+    operator std::filesystem::path() const { return make_path(); }
+    operator ss::sstring() const { return make_string(); }
+
+private:
+    // For segment_full_path::mock() in unit tests
+    partition_path(model::ntp ntp, model::revision_id r)
+      : ntp(ntp)
+      , revision_id(r) {}
+
+    std::filesystem::path make_path() const {
+        return std::filesystem::path(make_string());
+    }
+    ss::sstring make_string() const;
+    friend std::ostream& operator<<(std::ostream&, const partition_path&);
+    friend struct segment_full_path;
+};
+
+struct segment_full_path {
+    // This should always point to a compile-time string constant: we do not
+    // use any dynamic extensions, just simple thing like ".index" and
+    // ".log.staging"
+    std::string_view extension{".log"};
```

Comment:
if it is compile time constant maybe we can make it `constexpr` ?

Reply:
Went down a bit of a rabbit hole last night. It does seem like there are some fun tricks/hacks for requiring string literals. But a comment is probably good enough for this narrow/storage-internal use case.